classes/omnect_uboot_embedded_version.bbclass

```diff
@@ -81,15 +83,23 @@ python omnect_uboot_embed_version() {
     import os
     import shutil
     import gzip
+    import time
     from pathlib import Path
 
     # read the previously calculated version information
-    bootloader_version_file = d.getVar("DEPLOY_DIR_IMAGE") + "/omnect_bootloader_version"
-    try:
-        with open( bootloader_version_file, "r", encoding="utf-8") as f:
-            omnect_bootloader_version = f.read()
-    except OSError:
-        bb.fatal("Unable to read from bootloader version file \"%s\"" % (bootloader_version_file))
+    # NOTE:
+    #   yes, the following polling is rather ugly, but due to the nature of how
+    #   the checksum file is generated (done during recipe parsing) there is no
+    #   other possibility to busy-wait for the file to appear ...
+    while True:
+        bootloader_version_file = d.getVar("DEPLOY_DIR_IMAGE") + "/omnect_bootloader_version"
+        try:
+            with open( bootloader_version_file, "r", encoding="utf-8") as f:
+                omnect_bootloader_version = f.read()
+            break
+        except OSError:
+            bb.warn("Unable to read from bootloader version file \"%s\" wait for it to appear" % (bootloader_version_file))
+            time.sleep(3)
```

Comment:
I still have no explanation for the observed problems, I can only tell that I saw the build of bootloader-versioned failing without busy-looping the existence check.
I am also not convinced that the current version calculation is the right way given that the yocto manual states /wrt to DEPLOY_DIR_IMAGE:

"It must not be used directly in recipes when deploying files. Instead, it’s only useful when a recipe needs to “read” a file already deployed by a dependency. So, it should be filled with the contents of DEPLOYDIR by the deploy class or with the contents of IMGDEPLOYDIR by the image class."

I guess they have reason for that statement and who knows, maybe our problems stem from that ...
otoh, I also see the problem that the file is needed in several places (i.e. in several recipes), and deploying to the same file via DEPLOYDIR would possibly lead to other errors.

Reply:
we don't deploy the version-file from a task. it is generated on parse time of the recipe including `omnect_bootloader_versioning.bbclass`. we could use any dir and not `DEPLOY_DIR_IMAGE`. this shouldn't change anything. 

this is also, why i don't think that the change fixes the problem we experienced. at compile time the recipe including `omnect_bootloader_versioning.bbclass` was parsed multiple times before generating the artefact in task `do_compile()`. 

even if `omnect_bootloader_versioning.bbclass` is included from two recipes (e.g. different versions) matching `OMNECT_BOOTLOADER` the computed checksum should stay the same. 
a possible bug would be e.g.
`OMNECT_BOOTLOADER = "u-boot-imx"`
and u-boot-imx_1.bb and u-boot-imx_2.bb   have different `OMNECT_BOOTLOADER_CHECKSUM_FILES` content.
Maybe we should improve that `OMNECT_BOOTLOADER` doesn't contain `PN` only.

Reply:
apart from that all of the above is most likely right, I would like to understand how it could happen at all that the bootloader version file is not yet there in the image deploy directory when the bootloader-versioned recipe is about to get built: if the version file is generated during recipe parsing, shouldn't it always exist when bitbake starts to really build things?
also: can we be sure that the version calculation function is really run during parsing even in case none of the involved files inheriting the class changed nor did the class file itself?
I agree that the version file itself is most likely not the culprit for a wrong bootloader version in sw-description as well as in bootloader binary, because without use of DEPLOYDIR I can't see any caching happening. but I also have absolutely no idea, how an old version can be used at all, because from the sources the old version I don't see how this old version could be generated.